toolchain/check/call.cpp

```diff
@@ -37,6 +41,95 @@ enum class EntityKind : uint8_t {
 };
 }  // namespace
 
+// Performs overload resolution for a call to an overloaded function.
+// Uses clang to perform the overloading resolution.
+// Returns the resolved function, or `nullopt` if overload resolution failed.
+static auto PerformOverloadResolution(Context& context, SemIR::LocId loc_id,
+                                      SemIR::InstId callee_id,
+                                      llvm::ArrayRef<SemIR::InstId> arg_ids)
+    -> std::optional<SemIR::InstId> {
+  Diagnostics::AnnotationScope annotate_diagnostics(
+      &context.emitter(), [&](auto& builder) {
+        CARBON_DIAGNOSTIC(InCallToOverloadedCppFunction, Note,
+                          "in call to a Cpp (overloaded) function here");
+        builder.Note(loc_id, InCallToOverloadedCppFunction);
+      });
+
+  // Map Carbon call argument types to Cpp types.
+  llvm::SmallVector<clang::Expr*, 12> arg_exprs;
+  for (SemIR::InstId arg_id : arg_ids) {
+    auto arg_cpp_type = MapToCppType(context, arg_id);
+    if (!arg_cpp_type) {
+      CARBON_DIAGNOSTIC(CallArgTypeNotSupported, Error,
+                        "call arg of type {0} is not supported", TypeOfInstId);
+      context.emitter().Emit(loc_id, CallArgTypeNotSupported, arg_id);
+      return std::nullopt;
+    }
+    auto* arg_expr = new (context.ast_context()) clang::OpaqueValueExpr(
+        clang::SourceLocation(), arg_cpp_type->getNonReferenceType(),
+        clang::ExprValueKind::VK_LValue);
+    arg_exprs.emplace_back(arg_expr);
+  }
+
+  auto fn_type_inst =
+      context.types().GetAsInst(context.insts().Get(callee_id).type_id());
+  auto fn_type = fn_type_inst.TryAs<SemIR::OverloadedFunctionType>();
+  if (!fn_type) {
+    return std::nullopt;
+  }
+  auto overloaded_fn =
+      context.overloaded_functions().Get(fn_type->overloaded_function_id);
+
+  // add candidate functions
+  clang::OverloadCandidateSet candidate_set(
+      clang::SourceLocation(),
+      clang::OverloadCandidateSet::CandidateSetKind::CSK_Normal);
+
+  clang::ASTUnit* ast = context.sem_ir().cpp_ast();
+  CARBON_CHECK(ast);
+  clang::Sema& sema = ast->getSema();
+
+  sema.AddFunctionCandidates(overloaded_fn.candidate_functions, arg_exprs,
+                             candidate_set);
+
+  // find best viable function
+  clang::OverloadCandidateSet::iterator best_viable_fn;
+  clang::OverloadingResult overloading_result =
+      candidate_set.BestViableFunction(sema, clang::SourceLocation(),
+                                       best_viable_fn);
```

Comment:
It's probably worth clarifying why do we go through overload resolution when there's a single candidate.